packages/chain-mon/src/initialized-mon/service.ts

```diff
@@ -0,0 +1,434 @@
+}
+
+export class InitializedMonService extends BaseServiceV2<
+  InitializedMonOptions,
+  InitializedMonMetrics,
+  InitializedMonState
+> {
+  constructor(options?: Partial<InitializedMonOptions & StandardOptions>) {
+    super({
+      version,
+      name: 'initialized-mon',
+      loop: true,
+      options: {
+        loopIntervalMs: 1000,
+        ...options,
+      },
+      optionsSpec: {
+        rpc: {
+          validator: validators.provider,
+          desc: 'Provider for network to monitor balances on',
+        },
+        startBlockNumber: {
+          validator: validators.num,
+          default: -1,
+          desc: 'L1 block number to start checking from',
+          public: true,
+        },
+      },
+      metricsSpec: {
+        initializedCalls: {
+          type: Gauge,
+          desc: 'Successful transactions to tracked contracts emitting initialized event',
+          labels: ['label', 'address'],
+        },
+        unexpectedRpcErrors: {
+          type: Counter,
+          desc: 'Number of unexpected RPC errors',
+          labels: ['section', 'name'],
+        },
+      },
+    })
+  }
+
+  protected async init(): Promise<void> {
+    // Connect to L1.
+    await waitForProvider(this.options.rpc, {
+      logger: this.logger,
+      name: 'L1',
+    })
+
+    this.state.chainId = await getChainId(this.options.rpc)
+
+    const l1StartingBlockTag = networks[this.state.chainId].l1StartingBlockTag
+
+    if (this.options.startBlockNumber === -1) {
+      const block_number = l1StartingBlockTag != null ? (await this.options.rpc.getBlock(l1StartingBlockTag)).number : 0
+      this.state.highestUncheckedBlockNumber = block_number
+    } else {
+      this.state.highestUncheckedBlockNumber = this.options.startBlockNumber
+    }
+  }
+
+  protected async main(): Promise<void> {
+    if (
+      (await this.options.rpc.getBlockNumber()) <
+      this.state.highestUncheckedBlockNumber
+    ) {
+      this.logger.info('Waiting for new blocks')
+      return
+    }
+
+    const network = networks[this.state.chainId]
+    const contracts = network.contracts
+
+    const block = await this.options.rpc.getBlock(
+      this.state.highestUncheckedBlockNumber
+    )
+    this.logger.info('Checking block', {
+      number: block.number,
+    })
+
+    const transactions: TransactionResponse[] = []
+    for (const txHash of block.transactions) {
+      const t = await this.options.rpc.getTransaction(txHash)
+      transactions.push(t)
+    }
+
+    for (const transaction of transactions) {
+      for (const contract of contracts) {
+        const to = transaction.to != null ? transaction.to : transaction["creates"]
+        if (compareAddrs(contract.address, to)) {
+          try {
+            const transactionReceipt = await transaction.wait()
+            for (const log of transactionReceipt.logs) {
+              // keccak256("Initialized(suint8)") = 0x7f26b83ff96e1f2b6a682f133852f6798a09c465da95921460cefb3847402498
```

Comment:
I don't think that looking at the event will help because its possible to upgrade and not reinit, there will be more events, we would just need to manually know offchain what the values should be, a more reliable solution would be to use the storage layout and get the storage and then pull out the exact value that we need and ensure its set to 1, the old go `check-l2` script had this functionality

Reply:
ok will check out the old `check-l2` script. to make sure I understood correctly -- would the idea be checking when the value in the storage changes?

Reply:
just checked the FMA doc and the official action item is to "implement and maintain monitoring systems for early detection of failure modes 1. for `upgradeAndCall` 2. for `Initialized` events". you're right that its possible to upgrade and not reinit. thus, maybe we could additionally look for another event associated with `upgradeAndCall`?

Reply:
Yea the general idea here is we should have monitoring in place to detect unauthorized/unexpected upgrades and contract initialization calls

Reply:
@tynes Yea the general idea here is we should have monitoring in place to detect unauthorized/unexpected upgrades and contract initialization calls

Reply:
just added support for the `Upgraded` event